### PR TITLE
Define `Variable.__iter__`

### LINKED ISSUE #5443: 0-dim Variable converts to empty list
A 0-dim array should not be cast to a list.
cf. cupy/cupy#1432

### PR DESCRIPTION
**Pull request by toslunar**
Fix #5443.

### CODE REVIEW

File: chainer/variable.py [hunk 1]
```diff
@@ -589,6 +589,15 @@ def __len__(self):
         """
         return len(self.data)
 
+    def __iter__(self):
+        if self.ndim == 0:
+            raise TypeError('iteration over a 0-d array')
+        return self._iter()
```
Comment (kmaehashi) [inline chainer/variable.py]: (as discussed) it can be improved as 
```py
return (self[i] for i in six.moves.range(len(self)))
```

**Comment by toslunar** [comment]
I'm investigating on test failures (Caffe import/export).

File: chainer/variable.py [hunk 2]
```diff
@@ -724,6 +724,16 @@ def __len__(self):
         """
         return len(self.array)
 
+    def __iter__(self):
+        if chainer.config.enable_backprop:
+            warnings.warn(
+                'Backprop through `Variable.__iter__` is slow. Consider'
+                ' using `chainer.functions.split_axis`.',
+                RuntimeWarning)
```
Comment (toslunar) [inline chainer/variable.py]: Chainer could add `PerformanceWarning` like cupy/cupy#1607.

Comment (kmaehashi) [reply chainer/variable.py]: I think it is better, as there are some other functions that may want it (e.g. `chainer/functions/math/erf.py`)

File: chainer/variable.py [hunk 3]
```diff
@@ -724,6 +724,16 @@ def __len__(self):
...
+        if self.ndim == 0:
+            raise TypeError('iteration over a 0-d array')
```
Comment (kmaehashi) [inline chainer/variable.py]: How about do this check first, before emitting warning?

Comment (toslunar) [reply chainer/variable.py]: Done.

**Comment by kmaehashi** [comment]
#6617 has merged.

**Comment by toslunar** [comment]
Fixed the PR to use `PerformanceWarning`.

File: docs/source/upgrade.rst [hunk 4]
```diff
@@ -10,6 +10,13 @@ Most changes are carefully designed not to break existing code; however changes
 Chainer v6
 ==========
 
+``Variable`` Became a Subclass of ``collections.abc.Iterable``
+--------------------------------------------------------------
+
+Prior to Chainer v5, :class:`~chainer.Variable` did not implement `__iter__`.
+In Chainer v6, ``issubclass(chainer.Variable, collections.abc.Iterable)`` returns `True`.
+It is recommended to check types by ``isinstance(_, collections.abc.Sequence)`` to distinguish :class:`tuple` and :class:`list` from :class:`~chainer.Variable`.
```
Comment (toslunar) [inline docs/source/upgrade.rst]: From offline discussion:
- `collections.Iterable`, `collections.Sequence` in Python 2.
- Using `Sequence` is not always the best choice.

File: docs/source/upgrade.rst [hunk 5]
```diff
@@ -7,6 +7,17 @@ Upgrade Guide
...
 
+
+Chainer v7
+==========
+
+``Variable`` Became a Subclass of ``collections.abc.Iterable``
+--------------------------------------------------------------
+
+Prior to Chainer v6, :class:`~chainer.Variable` did not implement :meth:`~chainer.Variable.__iter__`.
+In Chainer v7, ``issubclass(chainer.Variable, collections.abc.Iterable)`` (or ``issubclass(chainer.Variable, collections.Iterable)`` with Python 2) returns `True`.
```
Comment (beam2d) [inline docs/source/upgrade.rst]: How about testing that `issubclass(Variable, Iterable)` is `True`?

Comment (toslunar) [reply docs/source/upgrade.rst]: Done.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
ping @kmaehashi

**Comment by kmaehashi** [comment]
Closing as Chainer is now under maintenance phase.

**Pull request closed without merging**